README.txt
Helpers for building asyncio-based apps
=========================================

Getting asyncio-based apps right can be a bit burdensome. Here's some helpers.

Please see the abstract base classes for the groundwork and the 'service' and 'wamp'
modules for the helper implementations and the tests on how to use them.

Note that to run the WAMP tests, you need to be able to run a crossbar WAMP router
locally. See the `test_wamp` module.


HISTORY.md

0.3rc1 (2016-12-08)
-------------------

- Misc. improvements and fixes

0.3rc0 (2016-12-07)
-------------------

- Add default waiter and runner that are restarted
  if they raise an Exception


tests/utils.py
import asyncio
import logging
import socketserver
import aiohttp
import aiohttp.server

from asynciohelpers.util import loggerprovider

from .protocols import MockTransportServerProtocol


def get_socket_server(loop, host, port):
   def factory():
      return MockTransportServerProtocol()
   return loop.create_server(factory, host, port)


def get_http_server(loop, host, port):
   def factory():
      return MockTransportServerProtocol()
   return loop.create_server(factory, host, port)


@loggerprovider
class TCPHandler(socketserver.BaseRequestHandler):

   def handle(self):
   # self.request is the TCP socket connected to the client
      while True:
         data = self.request.recv(1024).strip()
         if data:
            self._logger.debug("data received from socket: %s" % data)


class SingleHTTPRequestHandler(aiohttp.server.ServerHttpProtocol):

  async def handle_request(self, message, payload):
      response = aiohttp.Response(
          self.writer, 200, http_version=message.version
      )
      response.add_header('Content-Type', 'text/html')
      response.add_header('Content-Length', '18')
      response.send_headers()
      response.write(b'<h1>It Works!</h1>')
      await response.write_eof()

      # after serving a single request, shut down
      #self.shutdown()


tests/config.py
"""TEST CONFIGURATION"""

import logging

TEST_HOST = "127.0.0.1"
TEST_PORT = 8989

TEST_HTTP_HOST = "127.0.0.1"
TEST_HTTP_PORT = 8989

TEST_WAMP_HOST = "127.0.0.1"
TEST_WAMP_PORT = 8080

logging.basicConfig()
logger = logging.getLogger("testing")
LOGLEVEL = logging.DEBUG
logger.setLevel(LOGLEVEL)


tests/servers.py
"""
Server implementations that will be tested.
"""

import asyncio
from asynciohelpers.util import loggerprovider, wamp_configured
from asynciohelpers.service import AsyncioConnecting
from asynciohelpers.service import AsyncioReConnecting
from asynciohelpers.wamp import WAMPServiceMixin
from asynciohelpers.testing import LoggingServiceImpl, TransportClientProtocol
from .config import TEST_HTTP_HOST, TEST_WAMP_HOST, TEST_WAMP_PORT, TEST_HTTP_PORT, LOGLEVEL
from .components import WAMPComponent


@loggerprovider
class ConnectingAsyncioServer(AsyncioConnecting, LoggingServiceImpl):

   _host = TEST_HTTP_HOST
   _port = TEST_HTTP_PORT
   _transport_factory = TransportClientProtocol

   LOGLEVEL = LOGLEVEL


@loggerprovider
class ReConnectingAsyncioServer(AsyncioReConnecting, LoggingServiceImpl):

   _host = TEST_HTTP_HOST
   _port = TEST_HTTP_PORT
   _transport_factory = TransportClientProtocol

   LOGLEVEL = LOGLEVEL

   async def _run(self):
      self._logger.debug("runner running")

      while not self._closing:
         self._logger.debug("saying hello")
         try:
            self._transport.write(b"Hello world!")
         except Exception as exc:
            self._logger.warn("could not send: %s" % exc)
         await asyncio.sleep(2, loop=self._loop)

      self._logger.warn("runner stopped")


@loggerprovider
@wamp_configured
class ConnectingWAMPService(WAMPServiceMixin, AsyncioConnecting, LoggingServiceImpl):

   wmp_url = "ws://%s:%i/ws" % (TEST_WAMP_HOST, TEST_WAMP_PORT)
   wmp_realm = "realm1"
   wmp_sessioncomponent = WAMPComponent
   wmp_extra = None
   wmp_serializers = None

   LOGLEVEL = LOGLEVEL


@loggerprovider
@wamp_configured
class ReConnectingWAMPService(WAMPServiceMixin, AsyncioReConnecting, LoggingServiceImpl):

   wmp_url = "ws://%s:%i/ws" % (TEST_WAMP_HOST, TEST_WAMP_PORT)
   wmp_realm = "realm1"
   wmp_sessioncomponent = WAMPComponent
   wmp_extra = None
   wmp_serializers = None

   LOGLEVEL = LOGLEVEL


asynciohelpers/abcs.py
from abc import abstractmethod, abstractproperty, ABCMeta


class ServiceBaseABC(metaclass=ABCMeta):

   @abstractproperty
   def _host(self):
      "host or IP to connect to"

   @abstractproperty
   def _port(self):
      "port to connect to"

   @abstractproperty
   def _ssl(self):
      "whether or not to use SSL"

   @abstractmethod
   def start(self):
      "call to start the service"

   @abstractmethod
   def stop(self):
      "call to stop the service"

   @abstractmethod
   def _setup(self):
      "after start is called, do some setup here"

   @abstractmethod
   def _run(self):
      "actually run the service"

   @abstractmethod
   def _teardown(self):
      "perform cleanup here after stop is called"


class WAMPServiceABC(metaclass=ABCMeta):

   @abstractproperty
   def wmp_url():
      ""

   @abstractproperty
   def wmp_realm():
      ""

   # OPTIONAL, SO NOT PART OF ABC
   #@abstractproperty
   #def wmp_ssl():
   #   ""

   @abstractproperty
   def wmp_sessioncomponent():
      ""

   @abstractproperty
   def wmp_serializers():
      ""

   @abstractproperty
   def wmp_extra():
      ""

   @abstractproperty
   def _transport_factory():
      ""

   @abstractmethod
   def _component():
      ""


asynciohelpers/util.py
from abc import abstractproperty, abstractmethod
import os, logging, signal, types
from autobahn.websocket.util import parse_url


def env_configured(cls):
   "get _host, _port, and _ssl from CONNECT_HOST, CONNECT_PORT & CONNECT_SSL env vars"

   host = os.environ.get("CONNECT_HOST")
   if host:
      cls._host = host
   port = os.environ.get("CONNECT_PORT")
   if port:
      cls._port = port
   ssl = os.environ.get("CONNECT_SSL")
   if ssl:
      cls._ssl = ssl
   return cls


def wamp_configured(cls):
   "get host, port & ssl from wmp_url"

   (isSecureURL, host, port, resource, path, params) = parse_url(cls.wmp_url)

   cls._host = host
   cls._port = port

   if not getattr(cls, "wmp_ssl", None):
      cls._ssl = isSecureURL
   elif cls.wmp_ssl and not isSecureURL:
      raise RuntimeError(
         'the wmp_ssl class variable of %s conflicts with the "ws:" prefix '
         'of the wmp_url variable. Did you mean to use "wss:"?' % cls.__name__)
   else:
      cls._ssl = cls.wmp_ssl

   return cls


def wamp_env_configured(cls):
   "get wmp_url & wmp_realm from WAMP_ROUTER_URL & WAMP_REALM env vars"

   envvar = os.environ.get("WAMP_ROUTER_URL")
   if envvar:
      cls.wmp_url = envvar

   envvar = os.environ.get("WAMP_ROUTER_REALM")
   if envvar:
      cls.wmp_realm = envvar

   return cls


def signalled(signalvalue):
   "decorate a method to be called upon signal"

   def deco(f):
      signal.signal(signalvalue, f)
      return f

   return deco


def loggerprovider(cls):
   "bind the global logger to class, or create new using class name"
   cls._logger = logging.getLogger(cls.__name__)
   level = getattr(cls, "LOGLEVEL", None) or logging.INFO
   cls._logger.setLevel(level)
   return cls


def logged(*args):
   "decorate a method to log, optionally with a level"

   def wrapped(method, lvl):

      def wrapper(self, *args, **kwargs):
         self._logger.log(lvl, "%s called" % method.__name__)
         return method(self, *args, **kwargs)

      return wrapper

   if type(args[0]) == types.FunctionType:
      # no level given, args[0] is the decorated method
      return wrapped(args[0], logging.DEBUG)

   else:
      # args[0] has level
      def deco(method):
         return wrapped(method, args[0])

      return deco


def logmethod(name, *args, **kwargs):
   "just add a logging method call, optionally with a level"

   lvl = args[0] if args else logging.DEBUG
   async = True if kwargs.get("async") else False

   def deco(cls):
      "actual decorator"

      def wrapper(self, *args, **kwargs):
         self._logger.log(lvl, "%s called" % name)
         if async:
            yield

      loggermethod =  types.MethodType(wrapper, cls)
      setattr(cls, name, loggermethod)

      return cls

   return deco


def isabstractmethod(obj):
   is_function = isinstance(obj, types.FunctionType)
   has_name = obj.__class__.__name__ == "abstractmethod"
   return is_function and has_name


def provides_abc(klass, abc):
   "whether or not a class provides an abc"
   members = list(abc.__dict__.items())
   attrs = [k for k,v in members if isinstance(v, abstractproperty)]
   methods = [k for k,v in members if isabstractmethod(v)]
   for member in attrs + methods:
      if not any(member in base.__dict__ for base in klass.__mro__):
         return False
   return True




asynciohelpers/wamp.py
import asyncio
from autobahn.wamp.types import ComponentConfig
from autobahn.asyncio.websocket import WampWebSocketClientFactory
from autobahn.websocket.util import parse_url


class WAMPServiceMixin:
   "base mixin that provides WAMP configuration plus transport and component factory"

   async def _setup(self):
      await super()._setup()
      await self._protocol.factory._session_joined

   @property
   def _transport_factory(self):
      "create and return the transport factory"
      try:
         factory = WampWebSocketClientFactory(self._component, url=self.wmp_url, serializers=self.wmp_serializers, loop=self._loop)
      except Exception as exc:
         raise Exception("could not build transport factory: %s" % exc)
      else:
         factory._session_joined = asyncio.Future(loop=self._loop)
         self._logger.info("WAMP connecting to %s, realm '%s'" % (self.wmp_url, self.wmp_realm))
         return factory

   def _component(self):
      "component factory method that creates and returns the component"

      try:
         cfg = ComponentConfig(self.wmp_realm, self.wmp_extra)
      except Exception as exc:
         raise Exception("could not instantiate WAMP component config: %s" % str(exc))

      try:
         session = self.wmp_sessioncomponent(cfg)
      except Exception as exc:
         raise Exception("could not instantiate WAMP component: %s" % str(exc))
      else:
         session._loop = self._loop
         return session


   async def _wait(self):
      "default waiter"
      while True:
         await asyncio.sleep(900, self._loop)
         self._logger.info("waiting, still alive...")

   async def _run(self):
      "default runner"
      while True:
         await asyncio.sleep(900, self._loop)
         self._logger.info("running, still alive...")


asynciohelpers/service.py
from contextlib import suppress
import asyncio
from concurrent.futures import CancelledError
from .exceptions import SetupException


class AsyncioRunning:
   "base runner class that sets up and runs the loop & payload"

   RESTART_DELAY = 15 # seconds until waiter & runner are restarted

   # these three required per the ABC
   _host = None
   _port = None
   _ssl = None

   _loop = None
   _external_loop = False


   def set_loop(self, loop):
      self._loop = loop
      self._external_loop = True


   def _on_wait_completed(self, *args):
      "also stop the runner when waiting is complete"
      self._closing = True
      try:
         self._run_task.cancel()
      except:
         # it's possible the runnables have not been created yet...
         pass


   def start(self):

      self._loop = self._loop or asyncio.get_event_loop()
      self._logger.debug("%s start requested" % self.__class__.__name__)

      self._closing = False
      self._started = asyncio.Future(loop=self._loop)

      # start the runner coro after setup is done, or exit if error

      self._wait_task = self._loop.create_task(self._wait())
      self._wait_task.add_done_callback(self._on_wait_completed)

      def setup_complete(future):
         exc = future.exception()
         if exc:
            self._logger.error("setup failed, stopping immediately: %s" % str(exc))
            setup_failed = SetupException("failure: %s" % str(exc))
            self._started.set_exception(setup_failed)
         else:
            self._logger.debug("setup completed ok, scheduling runner")
            self._run_task = self._loop.create_task(self._run())
            self._started.set_result(True)

      self._setup_task = self._loop.create_task(self._setup())
      self._setup_task.add_done_callback(setup_complete)

      if self._external_loop:
         return self._started

      # run the setup
      try:
         self._loop.run_until_complete(self._setup_task)
      except Exception as exc:
         self._wait_task.cancel()
         self._loop.call_soon_threadsafe(self._loop.stop)

      # run the main loop; we can be stopped either by waiter completing & canceling
      # runner or by stop(), which actually just cancels the waiter

      while not self._closing:

         try:
            self._loop.run_until_complete(self._run_task)
         except CancelledError:
            self._logger.info("runner cancelled")
            break
         except KeyboardInterrupt:
            self._logger.info("runner terminated by user action")
            self._wait_task.cancel()
            break
         except Exception as exc:
            self._logger.error("runner failure: %s" % str(exc))
            self._delaying = asyncio.sleep(self.RESTART_DELAY, loop=self._loop)
            self._loop.run_until_complete(self._delaying)
            if not self._closing:
               self._run_task = self._loop.create_task(self._run())

      self._teardown_task = self._loop.create_task(self._teardown())

      try:
         self._loop.run_until_complete(self._teardown_task)
      except Exception as exc:
         self._logger.warn("teardown problem: %s" % exc)

      remaining = asyncio.Task.all_tasks()
      for task in remaining:
         task.cancel()

      try:
         self._loop.run_until_complete(asyncio.gather(*remaining))
      except CancelledError:
         pass
      except Exception as exc:
         self._logger.warn("cleanup failure: %s" % exc)
      try:
         self._loop.call_soon_threadsafe(self._loop.stop)
      except Exception as exc:
         self._logger.warn(str(exc))

      self._loop.close()
      self._logger.info("%s is now shut down" % self.__class__.__name__)


   def stop(self, *args, **kwargs):
      "cancel the task and return the stop future if external loop"

      self._logger.info("stop requested")
      self._wait_task.cancel()

      if self._external_loop:
         self._teardown_task = self._loop.create_task(self._teardown())
         return self._teardown_task



class AsyncioConnecting(AsyncioRunning):
   "asyncio service that connects a given transport"

   _transport_factory = None # protocol class, or other instance factory
   _host = None # FQDN
   _port = None # int
   _ssl = False # bool

   async def _connect(self):
      args = (self._transport_factory,)
      kwargs = {"host": self._host, "port": self._port, "ssl": self._ssl}
      connector = self._loop.create_connection(*args, **kwargs)
      (self._transport, self._protocol) = await connector
      self._protocol.is_closed = asyncio.Future(loop=self._loop)
      self._logger.debug("connected transport (%s)" % self._transport.__class__.__name__)

   async def _setup(self):
      self._logger.debug("connecting")
      await self._connect()

   async def _teardown(self):
      self._logger.debug("closing transport")
      try:
         self._transport.close()
      except Exception as exc:
         self._logger.warn("cannot close transport: %s" % exc)

      # protocol implementation MUST set this:
      await self._protocol.is_closed



class AsyncioReConnecting(AsyncioConnecting):
   "asyncio service that connects a given transport"

   RECONNECT_DELAY = 5 # seconds

   async def _connect(self):
      "after super() setup has run, register a protocol close (re)connect callback"

      await asyncio.sleep(self.RECONNECT_DELAY, loop=self._loop)

      while not self._closing:
         self._logger.debug("attempting connect")
         try:
            result = await super()._connect()
         except ConnectionError as exc:
            self._logger.warn("connection refused, retrying: %s" % str(exc))
         except CancelledError as exc:
            self._logger.error("connect cancelled, stopping: %s" % str(exc))
            break
         except SetupException as exc:
            self._logger.error("setup failed: %s" % exc)
            break
         except Exception as exc:
            self._logger.error("unhandled error %s: %s" % (type(exc), exc))
         else:
            self._protocol.is_closed.add_done_callback(self._reconnect)
            break
         await asyncio.sleep(self.RECONNECT_DELAY, loop=self._loop)

   def _reconnect(self, future):
      if not self._closing:
         self._logger.warn("connection lost, reconnecting")
         reconnect = self._loop.create_task(self._connect())
      else:
         self._logger.warn("already closing, not reconnecting")



asynciohelpers/testing.py
from abc import abstractmethod, abstractproperty, ABCMeta

import os, time, signal, logging
import subprocess
import asyncio
import logging
import socketserver
import aiohttp
import aiohttp.server

from contextlib import contextmanager
from asynciohelpers.util import loggerprovider
from .util import logmethod, loggerprovider, logged

LOGLEVEL = logging.DEBUG


class LoggingServiceImpl():
   "provide the implementables"

   async def _setup(self):
      self._logger.debug("%s setting up" % self.__class__.__name__)

   async def _run(self):
      self._logger.debug("%s runner running" % self.__class__.__name__)

   async def _wait(self):
      self._logger.debug("%s waiter started" % self.__class__.__name__)

   async def _teardown(self):
      self._logger.debug("%s tearing down" % self.__class__.__name__)


@logmethod("connection_lost")
@loggerprovider
class MockTransportServerProtocol(asyncio.Protocol):
   ""

   LOGLEVEL = LOGLEVEL

   @logged
   def connection_made(self, transport):
      self._transport = transport
      self._counter = 3

   def data_received(self, data):
      self._logger.debug("receiveing data from client: %s" % str(data))
      self._counter -= 1
      if self._counter == 0:
         self._logger.debug("will not receive any more, closing")
         self._transport.close()
      else:
         self._logger.debug("receiving %i more times" % self._counter)


def get_socket_server(loop, host, port):
   def factory():
      return MockTransportServerProtocol()
   return loop.create_server(factory, host, port)


def get_http_server(loop, host, port):
   def factory():
      return MockTransportServerProtocol()
   return loop.create_server(factory, host, port)


@logmethod("connection_lost")
@loggerprovider
class MockTransportServerProtocol(asyncio.Protocol):
   ""

   LOGLEVEL = LOGLEVEL

   @logged
   def connection_made(self, transport):
      self._transport = transport
      self._counter = 3

   def data_received(self, data):
      self._logger.debug("receiveing data from client: %s" % str(data))
      self._counter -= 1
      if self._counter == 0:
         self._logger.debug("will not receive any more, closing")
         self._transport.close()
      else:
         self._logger.debug("receiving %i more times" % self._counter)


@loggerprovider
class CloseNotifyingProtocol:

   LOGLEVEL = LOGLEVEL

   @logged
   def connection_lost(self, exc):
      self.is_closed.set_result(True)


@logmethod("connection_made")
@loggerprovider
class TransportServerProtocol(CloseNotifyingProtocol, asyncio.Protocol):

   LOGLEVEL = LOGLEVEL

   def connection_lost(self, exc):
      super().connection_lost(exc)



@loggerprovider
@logmethod("connection_made")
class TransportClientProtocol(CloseNotifyingProtocol, asyncio.Protocol):

   LOGLEVEL = LOGLEVEL

   def connection_lost(self, exc):
      super().connection_lost(exc)


@loggerprovider
class TCPHandler(socketserver.BaseRequestHandler):

   def handle(self):
   # self.request is the TCP socket connected to the client
      while True:
         data = self.request.recv(1024).strip()
         if data:
            self._logger.debug("data received from socket: %s" % data)


class SingleHTTPRequestHandler(aiohttp.server.ServerHttpProtocol):

  async def handle_request(self, message, payload):
      response = aiohttp.Response(
          self.writer, 200, http_version=message.version
      )
      response.add_header('Content-Type', 'text/html')
      response.add_header('Content-Length', '18')
      response.send_headers()
      response.write(b'<h1>It Works!</h1>')
      await response.write_eof()

      # after serving a single request, shut down
      #self.shutdown()


# context manager to start crossbar

@contextmanager
def crossbar_router():
   CBCMD = os.environ.get("CROSSBAR")
   assert CBCMD, "Must have environment variable CROSSBAR set to crossbar binary path"
   cdir = os.path.dirname(__file__)
   cdir = cdir[:cdir.rfind(os.path.sep)] + os.path.sep + "tests"
   cbp = subprocess.Popen([CBCMD, "start", "--cbdir", cdir], stdout=subprocess.DEVNULL)
   time.sleep(4)
   yield cbp
   cbp.terminate()
   cbp.wait()
   return


# asynchronous version

@loggerprovider
class CrossbarRouter:

   def __init__(self, loop=None):
      self.loop = loop or asyncio.get_event_loop()

   async def __aenter__(self):
      CBCMD = os.environ.get("CROSSBAR")
      assert CBCMD, "Must have environment variable CROSSBAR set to crossbar binary path"
      cdir = os.path.dirname(__file__)
      cdir = cdir[:cdir.rfind(os.path.sep)] + os.path.sep + "tests"
      coro = asyncio.create_subprocess_exec(CBCMD, "start", "--cbdir", cdir, stdout=subprocess.DEVNULL, loop=self.loop)
      self.cbp = await coro
      await asyncio.sleep(3)
      self._logger.debug("started WAMP router with pid %i" % self.cbp.pid)

   async def __aexit__(self, exc_type, exc, tb):
      self.cbp.terminate()
      await self.cbp.wait()
      await asyncio.sleep(1, loop=self.loop)


class NotImplementedABC(metaclass=ABCMeta):

   @abstractproperty
   def _not_implemented_property(self):
      ""

   @abstractmethod
   def _not_implemented_method(self):
      ""
